### DISCUSSION TITLE
What are components of Semantic kernel

### DISCUSSION THREAD

**Discussion by rakshahulle** [microsoft/semantic-kernel#9415] [Q&A] [answered]
The components of the Semantic kernel are not mentioned in an official document of the semantic kernel.

**Comment by moonbox3** [comment, marked as answer]
Hello @rakshahulle, have you taken a look at our [Microsoft Learn site documentation](https://learn.microsoft.com/en-us/semantic-kernel/overview/)?

**Comment by rakshahulle** [comment]
Yes

**Comment by rakshahulle** [comment]
Are concepts components? I have a dout

**Reply by moonbox3** [reply]
Yes, the concepts show the components. For example, the central [Kernel](https://learn.microsoft.com/en-us/semantic-kernel/concepts/kernel?pivots=programming-language-python) object where one specifies different [AI services](https://learn.microsoft.com/en-us/semantic-kernel/concepts/ai-services/) to interact with an LLM. One can bring their own native code as a [Plugin](https://learn.microsoft.com/en-us/semantic-kernel/concepts/plugins/?pivots=programming-language-python) or create a Plugin based on a prompt.

Is there a specific question you have that we can help answer?

**Comment by rakshahulle** [comment]
So, Kenel, AI service, Memory, Plugin, Planner , Prompt Engineering are
components of semantic kernel? Am I right? Also want to know about
Connectors, Semantic Function, Filters

**Reply by moonbox3** [reply]
I would say the highest level components are:

1. Plugins
2. Planners
3. Personas

Then Plugins:
- Plugin from a method (native code like C#, Python or Java)
- Plugin from a prompt (a natural language plugin to interact with an LLM)

Planners:
- Interacting with the model to satisfy a user query/goal
   - This can be done using one (or both) of the plugins from above.
   - One may use "auto function calling" to have the model grounded, as required by the developer
   - When interacting with the model, one can specify the components: Connectors (either memory or AI service like Azure OpenAI or any other available AI connector)

Personas:
- You can construct various personas based on your use-case/scenario. For example, I can create an "agent" to work a particular task. This could be a "travel agent" or a "reviewing agent." (This doesn't necessarily mean you use our Agent Framework, but that's possible, too.) As part of the persona one can bring in the plugins and planners aspect. I'm not specifically saying to use one of our Planners, but think about how to "plan" a task for the LLM agent like "Find me my next open slot on my calendar and book a lunch for me and my colleague." The LLM will break that down into a plan like: "Find the next open slot on Evan's calendar (using a plugin / auto function calling)" and then "using the result from the first plugin, book a lunch for Evan and his colleague (using another plugin / auto function calling)."

I'd say prompt engineer fits into the "plugin," "planner," or even the "persona" concept. It can take a bit to make sure one engineers the prompt effectively so they achieve their desired results. Same with filters -- filters provide an aspect to "hook" into certain places to the code -- like to view the final rendered prompt, or to get the function result from an auto function calling loop.

One last important one is Telemetry: we use OpenTelemetry to allow developers to get an end-to-end view of their Semantic Kernel agent or operation. One can view the prompt, the number of tokens used, and more.

**Comment by rakshahulle** [comment]
Thank you so much for your answer; it is so useful. But, I am now a
little bit confused with the component Persona. As there is no physical
component of persona(Is Agent persona) or Can We say Persona as Prompt.

**Reply by moonbox3** [reply]
You're welcome! I'm glad it helped.

Regarding the "persona" component: it’s more of a guiding concept than a physical entity. It defines the style, behavior, or focus of the AI Agent during interactions.

Think of it like a character or role that the agent adopts, shaping how it responds to users. While a prompt is a specific instruction, a persona is an ongoing set of preferences that influences how those prompts are interpreted.

So, a persona isn't a prompt itself but rather a guiding framework that makes the agent’s interactions feel consistent and tailored to user needs.

**Comment by rakshahulle** [comment]
Thanks a lot.
